lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs

```diff
@@ -12,1149 +12,324 @@
-    /// that may be requested in the future.
-    fn handle_format_list_response(&self, flags: ClipboardHeaderFlags) -> RdpResult<Messages> {
-        if !flags.contains(ClipboardHeaderFlags::CB_RESPONSE_OK) {
-            warn!("RDP server did not process our copy operation");
-        }
-        Ok(vec![])
+        self.requested_formats.push(format);
+        self.send("remote_copy", move |c| c.initiate_paste(format));
     }
 
-    /// Handles a request from the RDP server for clipboard data.
-    /// This message is received when a user executes a paste in the remote desktop.
-    fn handle_format_data_request(&self, payload: &mut Payload) -> RdpResult<Messages> {
-        let req = FormatDataRequestPDU::decode(payload)?;
-        let data = match self.clipboard.get(&req.format_id) {
-            Some(d) => d.clone(),
-            // TODO(zmb3): send empty FORMAT_DATA_RESPONSE with RESPONSE_FAIL flag set in header
+    fn on_format_data_request(&mut self, format: FormatDataRequest) {
+        trace!("CLIPRDR: on_format_data_request");
+        let response = match self.clipboard_data.as_ref() {
+            Some(data) => convert_string(&data, format.format),
             None => {
-                return Err(invalid_data_error(
-                    format!(
-                        "clipboard does not contain data for format {}",
-                        req.format_id
-                    )
-                    .as_str(),
-                ))
+                debug!(
+                    "format {:?} was requested but no data is available",
+                    format.format
+                );
+                None
             }
         };
-
-        self.add_headers_and_chunkify(
-            ClipboardPDUType::CB_FORMAT_DATA_RESPONSE,
-            FormatDataResponsePDU { data }.encode()?,
-        )
-    }
-
-    /// Receives clipboard data from the remote desktop. This is the server responding
-    /// to our format data request.
-    fn handle_format_data_response(
-        &mut self,
-        payload: &mut Payload,
-        length: u32,
-    ) -> RdpResult<Messages> {
-        let resp = FormatDataResponsePDU::decode(payload, length)?;
-        let data_len = resp.data.len();
-        let format = self.incoming_paste_formats.pop_front().ok_or_else(|| {
-            try_error("no expected format found, possibly received too many format data responses")
-        })?;
-
-        debug!(
-            "received {} bytes of copied data from Windows Desktop with format {:?}",
-            data_len, format,
-        );
-
-        let decoded = decode_clipboard(resp.data, format)?;
-        (self.on_remote_copy)(decoded)?;
-        Ok(vec![])
+        self.send("format_data_request", move |c| {
+            let response = match response.clone() {
+                None => FormatDataResponse::new_error(),
+                Some(data) => FormatDataResponse::new_data(data),
+            };
+            c.submit_format_data(response)
+        });
```

Comment:
nit: can remove unnecessary clone in the case of `None` by doing this like

```suggestion
        self.send("format_data_request", move |c| {
            let response = if let Some(data) = &response {
                FormatDataResponse::new_data(data.clone())
            } else {
                FormatDataResponse::new_error()
            };
            c.submit_format_data(response)
        });
```

Reply:
Does it give us something? `.clone()` for `None` is basically no-op